dali/kernels/slice/slice_hwc2chw_normalize_gpu.cu

```diff
@@ -0,0 +1,531 @@
+ */
+template <typename Out, typename In, bool enable_mirror, bool enable_pad, int kBlockSize,
+          int kStaticChannels>
+__global__ void Hwc2ChwNormalize(const Hwc2ChwSampleDesc<Out, In> *samples,
+                                 uint32_t num_samples) {
+  // TODO(klecki): generalize for wider input types
+  static_assert(std::is_same<In, uint8_t>::value, "Only uint8_t supported as input");
+
+  int sample_idx = FindSampleIdx(samples, num_samples);
+  const auto sample = samples[sample_idx];
+  int64_t start_x = (blockIdx.x - sample.first_block) * kBlockSize;
+  int64_t end_x = ::min(start_x + kBlockSize, sample.sample_size);
+  __shared__ float tile[kBlockSize + 32 * 4];
+
+  // Preload the norm values so they are accessed via registers and not from gmem via pointer.
+  float norm_mul[kStaticChannels], norm_add[kStaticChannels];
+
+  #pragma unroll kStaticChannels
+  for (int c = 0; c < kStaticChannels; c++) {
+    norm_mul[c] = sample.norm_mul[c];
+    norm_add[c] = sample.norm_add[c];
+  }
+
+  auto in_start = reinterpret_cast<std::uintptr_t>(sample.in + start_x);
+  auto aligned_in_start = align_up(in_start, 32 * 4);
+  auto bytes_skipped =
+      ::min(static_cast<int64_t>(aligned_in_start - in_start), end_x - start_x);
+
+  float *aligned_tile = tile + 32 * 4;
+  float *prologue_tile = aligned_tile - bytes_skipped;
+  const In *prologue_in = sample.in + start_x;
+
+  const uchar4 *aligned_in_char4 =
+      reinterpret_cast<const uchar4 *>(sample.in + start_x + bytes_skipped);
+
+  // prologue
+  for (int64_t idx = threadIdx.x; idx < bytes_skipped; idx += blockDim.x) {
+    prologue_tile[idx] = prologue_in[idx];
+  }
+  int64_t left_after_prologue = end_x - start_x - bytes_skipped;
+
+  // aligned load
+  for (int64_t idx = threadIdx.x; idx < left_after_prologue / 4; idx += blockDim.x) {
+    uchar4 in = aligned_in_char4[idx];
+    aligned_tile[idx * 4 + 0] = in.x;
+    aligned_tile[idx * 4 + 1] = in.y;
+    aligned_tile[idx * 4 + 2] = in.z;
+    aligned_tile[idx * 4 + 3] = in.w;
+  }
+  int64_t processed_in_main = (left_after_prologue / 4) * 4;
+  int64_t left_after_main = left_after_prologue - processed_in_main;
+
+  // epilogue
+  float *epilogue_tile = aligned_tile + processed_in_main;
+  const In *epilogue_in = reinterpret_cast<const In *>(aligned_in_char4 + processed_in_main / 4);
+
+  for (int64_t idx = threadIdx.x; idx < left_after_main; idx++) {
+    epilogue_tile[idx] = epilogue_in[idx];
+  }
+
+  __syncthreads();
+  const auto *__restrict__ fill_values = static_cast<const Out *>(sample.fill_values);
+
+  // idx is not divided by the static channels (mostly the start_x)
+  for (int64_t idx = threadIdx.x + start_x / kStaticChannels, base_x = threadIdx.x;
+       idx < end_x / kStaticChannels; idx += blockDim.x, base_x += blockDim.x) {
+    // TODO(klecki): forceinline device function
+    int64_t out_offset;
+    if constexpr (enable_mirror) {
```

Comment:
Does it help?

Reply:
Yes, not having those ifs if we do not use them actually helped.